ding/policy/wqmix.py

```diff
@@ -0,0 +1,455 @@
+
+        next_inputs = {'obs': data['next_obs']}
+        logit_detach = self._learn_model.forward(next_inputs, single_step=False, Q_star=False)['logit'].clone().detach()
+        next_inputs = {'obs': data['next_obs'], 'action': logit_detach.argmax(dim=-1)}
+
+        with torch.no_grad():
+            target_total_q = self._learn_model.forward(next_inputs, single_step=False, Q_star=True)['total_q']
+
+        with torch.no_grad():
+            if data['done'] is not None:
+                target_v = self._gamma * (1 - data['done']) * target_total_q + data['reward']
+            else:
+                target_v = self._gamma * target_total_q + data['reward']
+
+        td_error = (total_q - target_v).clone().detach()
+        _data = v_1step_td_data(total_q, target_total_q, data['reward'], data['done'], data['weight'])
+        _, td_error_per_sample = v_1step_td_error(_data, self._gamma)
+
+        data_star = v_1step_td_data(total_q_star, target_total_q, data['reward'], data['done'], data['weight'])
+        loss_star, td_error_per_sample_star_ = v_1step_td_error(data_star, self._gamma)
+
+        # our implemention is based on the https://github.com/oxwhirl/wqmix
+        # Weighting
+        alpha_to_use = self._cfg.learn.alpha
+        if self._cfg.learn.wqmix_ow:  # Optimistically-Weighted
+            ws = torch.ones_like(td_error) * alpha_to_use
+            # if td_error < 0, i.e. Q < y_i, then w =1; if not, w = alpha_to_use
+            ws = torch.where(td_error < 0, torch.ones_like(td_error) * 1, ws)
+        else:  # Centrally-Weighted
+            inputs = {'obs': data['obs']}
+            logit_detach = self._learn_model.forward(inputs, single_step=False, Q_star=False)['logit'].clone().detach()
+            cur_max_actions = logit_detach.argmax(dim=-1)
+            inputs = {'obs': data['obs'], 'action': cur_max_actions}
+            max_action_qtot = self._learn_model.forward(inputs, single_step=False, Q_star=True)['total_q']  # Q_star
+            # Only if the action of each agent is optimal, then the joint action is optimal
+            is_max_action = (data['action'] == cur_max_actions).min(dim=2)[0]  # shape (H,B,N) -> (H,B)
+            qtot_larger = target_v > max_action_qtot
+            ws = torch.ones_like(td_error) * alpha_to_use
+            # if y_i > Q_star or u =  u_star,  then w =1; if not, w = alpha_to_use
+            ws = torch.where(is_max_action | qtot_larger, torch.ones_like(td_error) * 1, ws)
+
+        if data['weight'] is None:
+            data['weight'] = torch.ones_like(data['reward'])
+        loss_weighted = (ws.detach() * td_error_per_sample * data['weight']).mean()
+
+        # ====================
+        # Q and Q_star update
+        # ====================
+        self._optimizer.zero_grad()
+        self._optimizer_star.zero_grad()
+        loss_weighted.backward(retain_graph=True)
+        loss_star.backward()
+        grad_norm = torch.nn.utils.clip_grad_norm_(
```

Comment:
You should name the return value of `clip_grad_norm_` separately for `q` and `q_star`, and visualize them separately too